src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -724,6 +757,128 @@ public PrefixEnumerable FindKeysWithPrefix(string prefix)
+
+            public void Copy(ModelStateEntry entry)
+            {
+                RawValue = entry.RawValue;
+                AttemptedValue = entry.AttemptedValue;
+                Errors.Clear();
+                for (var i = 0; i < entry.Errors.Count; i++)
+                {
+                    Errors.Add(entry.Errors[i]);
+                }
+
+                ValidationState = entry.ValidationState;
+                Visible = true;
+            }
+
+            public void Reset()
+            {
+                Visible = false;
+                RawValue = null;
+                AttemptedValue = null;
+                ValidationState = ModelValidationState.Unvalidated;
+                Errors.Clear();
+            }
+
+            public ModelStateNode GetNode(StringSegment subKey, bool createIfNotExists)
+            {
+                if (subKey.Length == 0)
+                {
+                    return this;
+                }
+
+                var index = BinarySearch(subKey);
+                ModelStateNode modelStateNode = null;
+                if (index >= 0)
+                {
+                    modelStateNode = ChildNodes[index];
+                }
+                else if (createIfNotExists)
+                {
+                    if (ChildNodes == null)
+                    {
+                        ChildNodes = new List<ModelStateNode>(1);
+                    }
+
+                    modelStateNode = new ModelStateNode(subKey);
+                    ChildNodes.Insert(~index, modelStateNode);
+                }
+
+                return modelStateNode;
+            }
+
+            public override ModelStateEntry GetModelStateForProperty(string propertyName)
+                => GetNode(new StringSegment(propertyName), createIfNotExists: false);
+
+            private int BinarySearch(StringSegment searchKey)
+            {
+                if (ChildNodes == null)
+                {
+                    return -1;
+                }
+
+                var low = 0;
+                var high = ChildNodes.Count - 1;
+                while (low <= high)
+                {
+                    var mid = low + ((high - low) / 2);
+                    var midKey = ChildNodes[mid].SubKey;
+                    int result;
+                    if (midKey.Length < searchKey.Length)
+                    {
+                        result = -1;
+                    }
+                    else if (midKey.Length > searchKey.Length)
+                    {
+                        result = 1;
+                    }
+                    else
+                    {
+                        result = string.Compare(
+                            midKey.Buffer,
+                            midKey.Offset,
+                            searchKey.Buffer,
+                            searchKey.Offset,
+                            searchKey.Length,
+                            StringComparison.OrdinalIgnoreCase);
+                    }
```

Comment:
In previous iteration, property `"AB"` would come before property `"Z"`. Not sure how much difference the intrinsic order makes but why place shorter property names first?